### PR TITLE
Remove more moment from state.

### PR DESCRIPTION
**Pull request by sgomes**
This removes a number of dates that were represented as moments in state, instead keeping them as their native API values.

This should remove the last bits of `i18n-calypso` `moment` usage.

#### Changes proposed in this Pull Request

* Remove moments from state
* Lint fixes
* Some other drive-by changes, such as switching from HOCs to hooks where it made sense

#### Testing instructions

I'm hoping the plans/purchases E2E tests cover most of the changed logic, but I've added the recommended reviewers from GitHub in hopes that they might be able to test further.

### CODE REVIEW

File: client/lib/purchases/index.js
Comment (sgomes) [inline client/lib/purchases/index.js]: @jsnajdr: I intend to fix this, but I'm not sure if it falls within the scope of this PR, or whether it's best handled in a separate one. What do you think?

File: client/lib/purchases/assembler.js
Comment (jsnajdr) [inline client/lib/purchases/assembler.js]: I'm not very happy about including derived values like this in the Redux state. Why not store just `mostRecentRenewDate` and let the consumers check the validity themselves?

If some computation is needed at multiple places, it can be a selector or a helper function.

File: client/lib/purchases/assembler.js
Comment (jsnajdr) [inline client/lib/purchases/assembler.js]: `expiryDateFormat` is not a credit-card-specific data, it's just a constant. Then why is it on the `creditCard` record?

And why is `expiryDateFormat` set to `null` on certain objects?

**Review by jsnajdr — COMMENTED** [review]
@sgomes would you appreciate if I pushed a few commits to this PR? I think many places in the code could be better, but I won't be sure until I actually try out the ideas 🙂

File: client/lib/purchases/assembler.js
Comment (sgomes) [reply client/lib/purchases/assembler.js]: `renewDate` has some specific semantics, and to centralise the validity checks, I I added `renewDateValid`. It seemed like more of a data layer concern rather than something that should be pushed on to consumers, which would need to all be aware of the semantics.

For consistency, I added it to the other ones as well with the simpler checks that match existing code.

File: client/lib/purchases/assembler.js
Comment (sgomes) [reply client/lib/purchases/assembler.js]: As I understand, there are two places where expiry dates appear: in the payment data, and in the credit card data. In the credit card data, they're always in the format of `MM/YY`. In the payment data, they can be either be in the default date format, or in `MM/YY` format in the case of PayPal Direct.

To try to make things consistent, I opted to add an `expiryDateFormat` field to both of them. It defaults to null (no specific format defined, i.e., just use the default date parsing), with `MM/YY` instead where applicable. When parsing with `moment`, you can simply do `moment( expiryDate, expiryDateFormat )` and have it parse the date according to the right format, or do something similar for other libraries.

This seemed to me like an extensible solution that makes it easy to change logic in the future if needed, without offloading that logic onto consumers.

**Comment by sgomes** [comment]
> @sgomes would you appreciate if I pushed a few commits to this PR? I think many places in the code could be better, but I won't be sure until I actually try out the ideas 🙂

Please, go right ahead! You tend to gave some good ideas 🙂

**Comment by sgomes** [comment]
@jsnajdr: I'm holding off on rebasing this to pick up the node 12 changes, in order to prevent any merge issue for you if you add any commits. Feel free to do so yourself if you'd like, however!

**Comment by sgomes** [comment]
Placing this "In Progress" due to changes introduced in #38486.

**Comment by sgomes** [comment]
Placed back in "Needs Review" after reworking atop the changes introduced in #38486.

@jsnajdr: I'm still doing things the exact same way, since I haven't yet come up with a better way of doing things, taking into account my replies to your comments. Happy to rework everything if you have any good suggestions, though!

**Comment by sgomes** [comment]
Note: `desktop` tests are failing in master, not as a consequence of this PR.

File: client/lib/purchases/assembler.js
Comment (jsnajdr) [reply client/lib/purchases/assembler.js]: It's true! If I look at a purchase I made with credits, the `expiry_date` is in ISO format. For credit card and PayPal, I didn't verify that, but it seems that `expiry_date` is `12/20`.

File: client/lib/purchases/index.js
Comment (jsnajdr) [inline client/lib/purchases/index.js]: If the expiry date is not valid, then this function returns `false` and we'll get a UI message like:
```
Your purchase is available until it expires on false
```
The original code would probably just display `null` instead of `false`, though.

Are there really any purchases that never expire?

File: client/me/purchases/manage-purchase/notices.jsx
Comment (jsnajdr) [inline client/me/purchases/manage-purchase/notices.jsx]: Here we call `fromNow()` on `false` if the expiry date is not valid. Again, the original code didn't check that either.

File: client/me/purchases/purchase-item/index.jsx
Comment (jsnajdr) [inline client/me/purchases/purchase-item/index.jsx]: I'm not sure if all these `expiryDateValid` and `expiry` checks are needed. They don't even seem to do a good job. If we display a message that says `Expires on false`, then what's the point? And it seems that the app worked just fine without them until now.

This is my main concern with this PR.

Comment (sgomes) [reply client/me/purchases/purchase-item/index.jsx]: Yeah, I can see your point. I'll go ahead and remove all of the `*Valid`, except for `renewDateValid`, since that one has special logic in place which I still think belongs in the "data" layer.

Comment (sgomes) [reply client/me/purchases/purchase-item/index.jsx]: Done.

File: client/me/purchases/manage-purchase/notices.jsx
Comment (sgomes) [reply client/me/purchases/manage-purchase/notices.jsx]: Check removed.

File: client/lib/purchases/index.js
Comment (sgomes) [reply client/lib/purchases/index.js]: I'm not sure if you can have purchases that don't expire. I've changed the code here, in any case.

**Review by ChaosExAnima — COMMENTED** [review]
Changes from #38486 look good here! I'll defer to others over other parts of the code :)

File: client/blocks/product-selector/index.jsx
Comment (tyxla) [inline client/blocks/product-selector/index.jsx]: Uber nit here: these props were ordered alphabetically before. Would be nice to maintain that.

File: client/blocks/product-selector/index.jsx
Comment (tyxla) [inline client/blocks/product-selector/index.jsx]: Just confirming that this is a good place for those `moment()` calls - the `ProductSelector` acts like a controller component.

File: client/lib/purchases/assembler.js
Comment (tyxla) [reply client/lib/purchases/assembler.js]: I came here to comment on the same thing 😉 

I wonder if it would make sense to make a decision and go with a single format in all cases. These moment objects that use that expiration field, after all, are being additionally formatted upon rendering of the date. Could using a single format be a good way to remove the unnecessary complexity here? Does it have any downsides?

File: client/lib/purchases/assembler.js
Comment (tyxla) [inline client/lib/purchases/assembler.js]: This feels suboptimal to me - having another field just to indicate whether a certain field is valid. I wonder if it makes sense to erase the `renewDate` field if it's not valid - the purchase assembler seems like a good place to do that.

File: client/me/purchases/purchase-item/index.jsx
Comment (tyxla) [reply client/me/purchases/purchase-item/index.jsx]: I think this still stands for the `expiry && ` checks though?

File: client/my-sites/checkout/checkout/recent-renewals.jsx
Comment (tyxla) [inline client/my-sites/checkout/checkout/recent-renewals.jsx]: Seems like `expiryString` isn't coming from anywhere right now. Did we forget to pass it down as a prop in `RecentRenewals`?

File: client/my-sites/checkout/checkout/recent-renewals.jsx
Comment (tyxla) [inline client/my-sites/checkout/checkout/recent-renewals.jsx]: I think we should clean up those checks - what is the reason to have them in the first place?

**Review by tyxla — COMMENTED** [review]
Thanks for working on this @sgomes! It's a great start.

I wonder if it makes sense to attack these fields one by one in separate PRs, as they seem to be used a lot throughout the app, and not in the most consistent way, so doing them one by one would help in assuring we're not breaking any of them inadvertently.

File: client/lib/purchases/assembler.js
Comment (jsnajdr) [reply client/lib/purchases/assembler.js]: > `renewDate` has some specific semantics

Some of these semantics seem to be just bugs that should be fixed in the server-side implementation of the REST endpoints. For example, if the renew date on a purchase is not set, its date string is `"0000-00-00 00:00:00"`. The PHP code parses and formats it with the `DateTime` PHP class, and the result sent in the REST response is `"-001-11-30T00:00:00+00:00"`. Note that month being set to zero means "December of the previous year", and similarly day set to zero means "last day of the previous month".

File: client/lib/purchases/assembler.js
Comment (jsnajdr) [reply client/lib/purchases/assembler.js]: Looking at the PHP code a bit closer, I think we can get rid of the special formatting strings.

The purchase itself can have three timestamps:
- `expiry_date`
- `subscribed_date`
- `renew_date`
All of them are either empty or always ISO strings. Some purchases never expire, e.g., if you buy a premium theme, it's yours forever.

Then there is a separate timestamp, `purchase_expiry_date` (transformed by the Calypso data layer to `purchase.payment.expiryDate`, which is present only on credit card and PayPal purchases, and this one is always in the `MM/YY` format.

**Comment by sgomes** [comment]
It sounds like we need to take a step back and make some decisions here, before proceeding further with this PR.

@tyxla @jsnajdr: It sounds like you're both in agreement that we should drop `*Valid` and instead fix the API, so that no special validation logic is required. This would mean that dates would always need to be returned in the same format and bugs like the one @jsnajdr described in https://github.com/Automattic/wp-calypso/pull/38467#discussion_r362733535 would need to be fixed.

That would lead to a significantly simpler PR here.

What do you think? Also, would anyone be able to help out with the PHP side of things? 😅

**Comment by jsnajdr** [comment]
> This would mean that dates would always need to be returned in the same format

I think that's already happening: dates in the `purchase` object are always ISO strings, and expiration months in `purchase.payment` are `MM/YY` strings (e.g., classic credit card expiration)

> Also, would anyone be able to help out with the PHP side of things?

I'll be working on this today 🙂

**Comment by jsnajdr** [comment]
Fixing the PHP dates in D37293-code. That makes the `renew_date > '0'` check redundant, as an empty date is always an empty string (`''`), never a weird negative date like `-001-11-30T00:00:00+00:00`.

**Comment by sgomes** [comment]
> I think that's already happening: dates in the `purchase` object are always ISO strings

As far as I understand, at least from reading the previous code, that's not true. PayPal Direct dates are returned in `purchase.payment_expiry`, and they come back in `MM/YY` format. See https://github.com/Automattic/wp-calypso/pull/38467/files#diff-4430bf8471f224634878b049437f0a60L100

**Comment by jsnajdr** [comment]
> Direct dates are returned in `purchase.payment_expiry`, and they come back in MM/YY format.

I'm talking about the `purchase` object that comes out of the `createPurchaseObject` function, not the original `purchase` object from the REST response.

REST field `purchase.payment_expiry` gets moved into the `purchase.payment` sub-object.

All dates that are in the top-level `purchase` are ISO strings.

**Comment by sgomes** [comment]
Rebased after @jsnajdr's massive linting fixes in #38142.

**Comment by jsnajdr** [comment]
@sgomes D37293-code wad deployed a few minutes ago. Complicated checks of the `renew_date` REST field should no longer be necessary. It's now either an empty string, or a valid ISO date string.

**Comment by sgomes** [comment]
> @sgomes D37293-code wad deployed a few minutes ago. Complicated checks of the `renew_date` REST field should no longer be necessary. It's now either an empty string, or a valid ISO date string.

Thank you, @jsnajdr! I'll start reworking this PR with that in mind.

File: client/blocks/product-selector/index.jsx
Comment (sgomes) [reply client/blocks/product-selector/index.jsx]: Fixed!

File: client/lib/purchases/assembler.js
Comment (sgomes) [reply client/lib/purchases/assembler.js]: Agreed. Now that the date format has been normalised (thank you to @jsnajdr for implementing and to you for reviewing!), we no longer need to do this 👍

File: client/me/purchases/purchase-item/index.jsx
Comment (sgomes) [reply client/me/purchases/purchase-item/index.jsx]: Removed!

File: client/my-sites/checkout/checkout/recent-renewals.jsx
Comment (sgomes) [reply client/my-sites/checkout/checkout/recent-renewals.jsx]: Good catch! This was an incomplete rename.

File: client/my-sites/checkout/checkout/recent-renewals.jsx
Comment (sgomes) [reply client/my-sites/checkout/checkout/recent-renewals.jsx]: My concern was that the `moment()` call could return a falsy value depending on the values provided, and thus this would generate an error accessing the property. After looking at the documentation, this doesn't appear to be possible.

File: client/lib/purchases/assembler.js
Comment (sgomes) [reply client/lib/purchases/assembler.js]: I think I finally fully grokked this. There's actually three (possible) expiry dates in the Calypso data layer object. There's `purchase.expiryDate`, which is always empty/ISO string; there's `purchase.payment.expiryDate`, which is only set in the case of PayPal Direct and is always `MM/YY`; and there's `purchase.payment.creditCard.expiryDate`, which is only set in the case of credit card and is always `MM/YY`.

The code has been updated to remove `expiryDateFormat` in light of these fixed formats.

**Comment by sgomes** [comment]
The PR has now been massively simplified, removing both `*Valid` and `*Format` throughout. Thank you both, @jsnajdr and @tyxla! Your suggestions and contributions have improved this massively 🙇 

Ready for another review round!

File: client/lib/purchases/assembler.js
Comment (jsnajdr) [inline client/lib/purchases/assembler.js]: Here, to avoid recreating the large `object`, I would simply assign to `object.payment.creditCard`:
```js
if ( isCreditCard ) {
  object.payment.creditCard = ...;
}
```
That should work equally well.

File: client/state/sites/plans/selectors.js
Comment (jsnajdr) [inline client/state/sites/plans/selectors.js]: This logic is difficult to understand and possibly buggy. If there is no current plan or if it doesn't have `expiryDate` set, what is `isCurrentPlanExpiring` supposed to return? Currently it seems to be `true`, because `null < 1234` is `true`. Is that really desired?

We should be explicit about that:
```js
if ( ! currentPlan || ! currentPlan.expiryDate ) {
  return true; // or false 😊
}
```

File: client/lib/purchases/assembler.js
Comment (sgomes) [reply client/lib/purchases/assembler.js]: Good idea, thank you!

File: client/lib/purchases/index.js
Comment (jsnajdr) [inline client/lib/purchases/index.js]: This is another place that should be more explicit about what's the return value when `purchase.expiryDate` is empty.

File: client/lib/purchases/index.js
Comment (jsnajdr) [inline client/lib/purchases/index.js]: This code is guarded by `isPaidWithCreditCard` and `hasCreditCardData`. If these checks are successful, it more or less means that `purchase.payment.creditCard.expiryDate` is set and valid. The extra checks are not necessary.

Also, the `purchase.expiryDate` string can be passed directly as argument of `diff`:
```js
moment( card.expiryDate ).diff( purchase.expiryDate, 'months' ) < 0
```
The parameter has type `MomentInput`.

An alternative (and better IMO) way to write the condition is:
```js
moment( creditCard.expiryDate, 'MM/YY' ).isBefore( purchase.expiryDate, 'months' )
```

File: client/state/sites/plans/selectors.js
Comment (sgomes) [reply client/state/sites/plans/selectors.js]: It's not clear to me either what it *should* be, so it's probably best to maintain the current behaviour. But I completely agree that it should be more explicit, so I went with your suggestion.

File: client/lib/purchases/index.js
Comment (sgomes) [reply client/lib/purchases/index.js]: Indeed. Fixed!

**Review by jsnajdr — APPROVED** [review]
I think this is ready to :ship: now. Thanks @sgomes for your hard work here!

**Comment by sgomes** [comment]
> I think this is ready to 🚢 now. Thanks @sgomes for your hard work here!

Woohoo! Thank you for all your contributions in getting the server side of things to play along, and for all your great review suggestions! 🙂

**Comment by tyxla** [comment]
Wonderful work on this one 🚀

**Comment by sgomes** [comment]
Thank you so much for all your review help as well, @tyxla, both on this PR and the server-side work!

**Comment by DavidRothstein** [comment]
I did some testing, and all the dates seem to be displaying the same as before on the payments pages.  Nice work!

**Pull request merged**
